File: app/src/test/java/no/nav/tilbakemeldingsmottak/validators/MeldFeilOgManglerValidatorTest.java
package no.nav.tilbakemeldingsmottak.validators;

import no.nav.tilbakemeldingsmottak.exceptions.ClientErrorException;
import no.nav.tilbakemeldingsmottak.model.MeldFeilOgManglerRequest;
import no.nav.tilbakemeldingsmottak.rest.feilogmangler.validation.MeldFeilOgManglerValidator;
import org.junit.jupiter.api.Test;

import static no.nav.tilbakemeldingsmottak.TestUtils.createMeldFeilOgManglerRequest;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class MeldFeilOgManglerValidatorTest {

    private final MeldFeilOgManglerValidator meldFeilOgManglerValidator = new MeldFeilOgManglerValidator();
    private MeldFeilOgManglerRequest meldFeilOgManglerRequest;

    @Test
    void happyPath() {
        meldFeilOgManglerRequest = createMeldFeilOgManglerRequest();
        meldFeilOgManglerValidator.validateRequest(meldFeilOgManglerRequest);
    }

    @Test
    void shouldThrowExceptionIfOnskerKontaktTrueAndEpostNotSet() {
        meldFeilOgManglerRequest = createMeldFeilOgManglerRequest();
        meldFeilOgManglerRequest.setOnskerKontakt(true);
        meldFeilOgManglerRequest.setEpost(null);
        Exception thrown = assertThrows(ClientErrorException.class,
                () -> meldFeilOgManglerValidator.validateRequest(meldFeilOgManglerRequest));
        assertTrue(thrown.getMessage().contains("epost er påkrevd dersom onskerKontakt=true"));
    }

    @Test
    void shouldThrowExceptionIfFeiltypeNotSet() {
        meldFeilOgManglerRequest = createMeldFeilOgManglerRequest();
        meldFeilOgManglerRequest.setFeiltype(null);
        Exception thrown = assertThrows(ClientErrorException.class,
                () -> meldFeilOgManglerValidator.validateRequest(meldFeilOgManglerRequest));
        assertTrue(thrown.getMessage().contains("feiltype er påkrevd"));
    }

    @Test
    void shouldThrowExceptionIfMeldingNotSet() {
        meldFeilOgManglerRequest = createMeldFeilOgManglerRequest();
        meldFeilOgManglerRequest.setMelding(null);
        Exception thrown = assertThrows(ClientErrorException.class,
                () -> meldFeilOgManglerValidator.validateRequest(meldFeilOgManglerRequest));
        assertTrue(thrown.getMessage().contains("melding er påkrevd"));
    }

}

File: app/src/test/kotlin/no/nav/tilbakemeldingsmottak/validators/MeldFeilOgManglerValidatorTest.kt
package no.nav.tilbakemeldingsmottak.validators

import no.nav.tilbakemeldingsmottak.exceptions.ClientErrorException
import no.nav.tilbakemeldingsmottak.model.MeldFeilOgManglerRequest
import no.nav.tilbakemeldingsmottak.rest.feilogmangler.validation.MeldFeilOgManglerValidator
import no.nav.tilbakemeldingsmottak.util.builders.MeldFeilOgManglerRequestBuilder
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

internal class MeldFeilOgManglerValidatorTest {
    private val meldFeilOgManglerValidator = MeldFeilOgManglerValidator()
    private var meldFeilOgManglerRequest: MeldFeilOgManglerRequest = MeldFeilOgManglerRequestBuilder().build()

    @Test
    fun happyPath() {
        meldFeilOgManglerRequest = MeldFeilOgManglerRequestBuilder().build()
        meldFeilOgManglerValidator.validateRequest(meldFeilOgManglerRequest)
    }

    @Test
    fun shouldThrowExceptionIfOnskerKontaktTrueAndEpostNotSet() {
        // Given
        meldFeilOgManglerRequest = MeldFeilOgManglerRequestBuilder().build(onskerKontakt = true, epost = null)

        // When
        val thrown: Exception = Assertions.assertThrows(ClientErrorException::class.java) {
            meldFeilOgManglerValidator.validateRequest(meldFeilOgManglerRequest)
        }

        // Then
        assertTrue(thrown.message!!.contains("epost er påkrevd dersom onskerKontakt=true"))
    }

    @Test
    fun shouldThrowExceptionIfFeiltypeNotSet() {
        // Given
        meldFeilOgManglerRequest = MeldFeilOgManglerRequestBuilder().build(feiltype = null)

        // When
        val thrown: Exception = Assertions.assertThrows(ClientErrorException::class.java) {
            meldFeilOgManglerValidator.validateRequest(meldFeilOgManglerRequest)
        }

        // Then
        assertTrue(thrown.message!!.contains("feiltype er påkrevd"))
    }

    @Test
    fun shouldThrowExceptionIfMeldingNotSet() {
        // Given
        meldFeilOgManglerRequest = MeldFeilOgManglerRequestBuilder().build(melding = null)

        // When
        val thrown: Exception = Assertions.assertThrows(ClientErrorException::class.java) {
            meldFeilOgManglerValidator.validateRequest(meldFeilOgManglerRequest)
        }

        // Then
        assertTrue(thrown.message!!.contains("melding er påkrevd"))
    }
}